src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs

```diff
@@ -2174,43 +2198,18 @@ private AssemblySymbol GetForwardedToAssembly(string fullName, int arity, Diagno
                 while (current != null);
             }
 
-            // NOTE: This won't work if the type isn't using CLS-style generic naming (i.e. `arity), but this code is
-            // only intended to improve diagnostic messages, so false negatives in corner cases aren't a big deal.
-            var metadataName = MetadataTypeName.FromFullName(fullName, useCLSCompliantNameArityEncoding: true, forcedArity: arity);
-
-            var containingAssembly = this.Compilation.Assembly;
+            name = MetadataHelpers.ComposeAritySuffixedMetadataName(name, arity);
 
-            // This method is only called after lookup has failed, so the containing (source!) assembly can't
-            // have a forwarder to another assembly.
-            NamedTypeSymbol forwardedType = null;
-            foreach (var referencedAssembly in containingAssembly.Modules[0].GetReferencedAssemblySymbols())
+            var fullName = qualifierOpt != null && !ReferenceEquals(qualifierOpt, Compilation.GlobalNamespace) ? qualifierOpt.ToDisplayString(SymbolDisplayFormat.QualifiedNameOnlyFormat) + "." + name : name;
+            var result = GetForwardedToAssembly(fullName, diagnostics, location);
+            if ((object)result != null || diagnostics.HasAnyErrors())
```

Comment:
>diagnostics [](start = 42, length = 11)

This appears to assume that the incoming diagnostic bag is empty. Please don't do that. The caller is welcome to keep using the same diagnostic bag to let reportable errors accumulate. The behavior of this code should not depend on whether or not the incoming diagnostic bag is empty.